### PR TITLE
feat(s3): add support for custom S3-compatible endpoints

### PR DESCRIPTION
**Pull request by njzc**
Allow users to specify a custom endpoint URL in the AWS S3 settings. This enables compatibility with S3-compatible storage providers (e.g., MinIO, DigitalOcean Spaces) by setting the endpoint and enabling path-style access.

- Update `AwsS3Setting` interface to include `endpoint`
- Add endpoint configuration field to `PublishSettingTab` UI
- Update `AwsS3Uploader` to use the custom endpoint if provided
- Add unit tests for `AwsS3Setting` configuration

### CODE REVIEW

**Review by addozhang — CHANGES_REQUESTED** [review]
Thanks for the PR! The feature makes sense — S3-compatible endpoint support is a common need. A few things to address before merging:

1. **Trim the endpoint value** — `setting.endpoint` should be trimmed before use. A whitespace-only or trailing-newline value is truthy but will produce an invalid endpoint and incorrectly enable `s3ForcePathStyle`. Suggest:
   ```ts
   const endpoint = setting.endpoint?.trim();
   if (endpoint) {
     s3Options.endpoint = endpoint;
     s3Options.s3ForcePathStyle = true;
   }
   ```

2. **Tests need more coverage** — The current tests only verify the interface shape, which TypeScript already guarantees at compile time. Please add tests that verify the S3 client is actually configured correctly when `endpoint` is provided (e.g., mock `AWS.S3` and assert the constructor args include `endpoint` and `s3ForcePathStyle: true`).

3. **Default region change** — The region fallback was changed from no default to `"us-east-1"`. While minor, this is a behavior change for existing users who rely on region being explicitly set. Please either revert this or call it out as intentional.

**Comment by addozhang** [comment]
One more thing:

4. **Inconsistent optional chaining** — In `publishSettingTab.ts`, `.setValue()` uses `this.plugin.settings.awsS3Setting?.endpoint` with optional chaining, but `.onChange()` directly accesses `this.plugin.settings.awsS3Setting.endpoint` without it. Since this code is already inside the S3 settings panel, `awsS3Setting` won't be null — please remove the unnecessary `?.` for consistency.

File: src/uploader/s3/awsS3Uploader.ts [hunk 1]
```diff
@@ -29,13 +37,15 @@ export default class AwsS3Uploader implements ImageUploader {
...
+    path = path.replace(/^\/+/, '');
     await this.s3.send(new PutObjectCommand({
       Bucket: this.bucket,
       Key: path,
       Body: uint8Array,
     }));
-    const location = `https://${this.bucket}.s3.${this.region}.amazonaws.com/${path}`;
+    const location = this.region
+      ? `https://${this.bucket}.s3.${this.region}.amazonaws.com/${path}`
+      : path;
```
Comment (addozhang) [inline src/uploader/s3/awsS3Uploader.ts]: **Bug: wrong URL returned when using a custom endpoint**

When `endpoint` is set but `region` is empty, `location` is set to just `path` (e.g. `images/foo.png`) — a relative path, not a usable URL. `customizeDomainName()` only replaces the host portion of an absolute URL, so if no `customDomainName` is configured the caller receives a bare path instead of an accessible object URL.

With `forcePathStyle` the correct base URL is `${endpoint}/${bucket}/${path}`. Suggested fix:

```typescript
let location: string;
if (this.customDomainName) {
  // customizeDomainName will substitute the host; build a canonical URL to rewrite
  location = `https://${this.bucket}.s3.amazonaws.com/${path}`;
} else if (this.endpoint) {
  location = `${this.endpoint}/${this.bucket}/${path}`;
} else {
  location = `https://${this.bucket}.s3.${this.region}.amazonaws.com/${path}`;
}
return UploaderUtils.customizeDomainName(location, this.customDomainName);
```

This requires storing `endpoint` as an instance variable (alongside the existing `region`, `bucket`, etc.).

File: tests/unit/s3Uploader.test.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+      customDomainName: "",
+      endpoint: "",
+    };
+
+    new AwsS3Uploader(setting);
+
+    const callArgs = mockS3Client.getLastCallArgs();
+    expect(callArgs.region).toBe("eu-west-1");
+  });
+});
```
Comment (addozhang) [inline tests/unit/s3Uploader.test.ts:238]: **Missing test: URL generation with custom endpoint**

The test suite covers `S3Client` constructor arguments well, but there is no test for the `upload()` return value when `endpoint` is set. Given the URL construction bug above, please add a test that mocks `PutObjectCommand` + `send()` and asserts the returned URL is `${endpoint}/${bucket}/${path}` (path-style) rather than just `path` or an AWS default URL.

**Review by addozhang — CHANGES_REQUESTED** [review]
The contributor has addressed all the previous feedback (trimming endpoint, deep-merging `awsS3Setting`, and replacing the shape-only tests with proper `vi.mock`-based constructor tests). Good work overall.

One blocking issue remains before merging:

**URL construction is broken when using a custom endpoint.**  
`awsS3Uploader.ts` line 48 falls back to returning the bare `path` string (e.g. `images/foo.png`) when `region` is empty. This happens whenever a user sets a custom endpoint but leaves `region` blank. The returned value is not a valid URL and the image reference in the document will be broken.

See the inline comment for a suggested fix and the corresponding test coverage request.

File: src/uploader/s3/awsS3Uploader.ts [hunk 3]
```diff
@@ -6,21 +6,31 @@ export default class AwsS3Uploader implements ImageUploader {
...
   private readonly region: string;
+  private readonly endpoint: string;
   private pathTmpl: string;
   private customDomainName: string;
 
 
   constructor(setting: AwsS3Setting) {
-    const region = UploaderUtils.trimCredential(setting.region);
-    this.s3 = new S3Client({
+    const region = UploaderUtils.trimCredential(setting.region) || (setting.endpoint?.trim() ? "us-east-1" : "");
```
Comment (addozhang) [inline src/uploader/s3/awsS3Uploader.ts]: The latest fix resolves the bare-path URL bug, but this should use `UploaderUtils.normalizeEndpoint(setting.endpoint)` instead of `setting.endpoint?.trim()`.

`normalizeEndpoint()` already exists and strips trailing slashes in addition to whitespace. Without it, a common input like `https://minio.example.com/` is stored as-is and `upload()` returns `https://minio.example.com//bucket/key` because line 52 appends `/${bucket}/${path}`.

Suggested change:

```typescript
const endpoint = UploaderUtils.normalizeEndpoint(setting.endpoint);
```

Please also add a URL-generation test for an endpoint with a trailing slash.

**Review by addozhang — CHANGES_REQUESTED** [review]
The previous blocking issue is fixed: custom endpoints now return path-style URLs and the new upload tests cover that behavior.

One remaining edge case should be fixed before merging: normalize the endpoint with the existing `UploaderUtils.normalizeEndpoint()` helper. Otherwise endpoint values with trailing slashes produce double-slash object URLs.
